[WIP]Add a LatticeModelProblem class

a-matsuo opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
Implemented a `LatticeModelProblem` class.

[] get_default_filter_criterion
[] `__str__` of `LatticeModelResult`
[] `num_occupied_modals_per_mode` of `LatticeModelResult`

### Details and comments
resolve #563

a-matsuo commented:
It is still messy, but at least it works with `NumPyMinimumEigensolver` like the following code.
```
lmp = LatticeModelProblem(lattice_model)
np_solver = NumPyMinimumEigensolverFactory()
calc = GroundStateEigensolver(qubit_converter, np_solver)
res = calc.solve(lmp)
```

I have questions for the following method. I couldn't find the corresponding codes for them. Can I skip them? I think we need `__str__` of `LatticeModelResult` at least.
- [ ]  get_default_filter_criterion
- [ ]  `__str__` of `LatticeModelResult`
- [ ] `num_occupied_modals_per_mode` of `LatticeModelResult`
@pbark @ikkoham

ikkoham commented:
- I think `get_default_filter_criterion` can return `None` in `LatticeModelProblem`.
- `__str__` returns only energy, but it ie better to have header like `=== GROUND STATE ENERGY ===`. In the future, we can extend to have magnetization, or some observables...
- we don't need `num_occupied_modals_per_mode` and ElectronicStructureResult  and EigenstateResult don't have it.

woodsp-ibm commented on qiskit_nature/problems/second_quantization/lattice/lattice_model_problem.py:
```diff
@@ -0,0 +1,110 @@
+        """
+        return [self._lattice_model.second_q_ops()]
+
+    def interpret(
```
Can we do some unit test(s) that cover the interpret and the result. As far as I can see these aspects are not covered at all.

a-matsuo commented:
@woodsp-ibm Thank you for reviewing. I added unittests for `interpret`. I will add some brief explanation how to use LatticeModelProblem in the tutorial.

a-matsuo commented on qiskit_nature/problems/second_quantization/lattice/lattice_model_problem.py:
```diff
@@ -0,0 +1,110 @@
+        Returns:
+            None
+        """
+        return None
```
Actually, I have no idea about it (I don't even know what qeom is) . But, I changed it to return `NotImplementedError` for the future.

a-matsuo commented:
@mrossinek @woodsp-ibm Thank you for reviewing. I fixed the code based on the comments. I also added some explanation for the Lattice Model Problem in the tutorial.

mrossinek commented on docs/tutorials/10_lattice_models.ipynb:
```diff
@@ -813,6 +813,14 @@
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "<ipython-input-21-3bf38577149b>:1: DeprecationWarning: The reduce function is deprecated as of version 0.4.0 and will be removed no sooner than 3 months after the release Instead, use `simplify` and optionally `normal_ordered`..\n",
+      "  ham = fhm.second_q_ops(display_format=\"sparse\").reduce()\n"
```
Could you change this to use `simplify()` instead?

mrossinek commented on docs/tutorials/10_lattice_models.ipynb:
```diff
@@ -894,50 +902,99 @@
+   "metadata": {},
+   "source": [
+    "## LatticeModelProblem\n",
+    "We also have the Lattice Model Problem to generate the Second quantized operator from a Lattice Model. We can use the Ground State Eigensolver with it to calculate the ground state energy."
```
```suggestion
    "Qiskit Nature also has a `LatticeModelProblem` class which allows the usage of the `GroundStateEigensolver` to calculate the ground state energy of a given lattice. You can use this class as follows:"
```

mrossinek commented on docs/tutorials/10_lattice_models.ipynb:
```diff
@@ -894,50 +902,99 @@
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/Users/atsushi/github/nature/qiskit_nature/problems/second_quantization/lattice/models/fermi_hubbard_model.py:123: UserWarning: The default value for `display_format` will be changed from 'dense' to 'sparse' in version 0.3.0. Once that happens, you must specify display_format='dense' directly.\n",
+      "  return FermionicOp(ham, register_length=register_length, display_format=display_format)\n"
```
Could you go into the `FermiHubbardModel` and explicitly set `display_format="sparse"` in order to remove this warning?

mrossinek commented on docs/tutorials/10_lattice_models.ipynb:
```diff
@@ -894,50 +902,99 @@
-     },
-     "metadata": {},
-     "output_type": "display_data"
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "=== GROUND STATE ===\n",
+      " \n",
+      "* Lattice ground state energy : -2.566350190841\n"
```
I am not sure, if this is because of the Github plain-text view of the tutorial, but I don't think this output cell should be here, but rather should be below the next cell

mrossinek commented on docs/tutorials/10_lattice_models.ipynb:
```diff
@@ -894,50 +902,99 @@
+   "source": [
+    "from qiskit.algorithms import NumPyMinimumEigensolver\n",
+    "from qiskit_nature.algorithms import GroundStateEigensolver\n",
+    "from qiskit_nature.mappers.second_quantization import JordanWignerMapper\n",
```
If I recall correctly, the type of lattice dictates what mapper to use, correctly? Is this being discussed anywhere in this tutorial? If not, this could be an addition for the future (but there is no direct need to add this now as part of this PR)

a-matsuo replied:
hmm, I have not mentioned about it in this tutorial (I didn't even know that). I will leave it for the future.

mrossinek commented on releasenotes/notes/lattice_model_problem-3925bfbd2267cd4c.yaml:
```diff
@@ -0,0 +1,12 @@
+    We can use the Ground State Eigensolver with it to calculate the ground state energy.
+
+    .. code-block:: python
+
+        calc = GroundStateEigensolver(converter, solver)
+        res = calc.solve(LatticeModelProblem(lattice_model))
```
I think, if anything, this should be a fully functional snippet with all imports and variable definitions properly working. You can copy the cell from the tutorial if you want.

ikkoham reviewed (commented):
Totally good direction. I have a very minor comment.

mrossinek commented on qiskit_nature/problems/second_quantization/lattice/lattice_model_problem.py:
```diff
@@ -0,0 +1,116 @@
+class LatticeModelProblem(BaseProblem):
+    """Lattice Model Problem class to create second quantized operators from a lattice model."""
+
+    def __init__(self, lattice_model=LatticeModel) -> None:
```
I don't think I fully understand this init argument. Did you want to specify the type here but
accidentally assigned the default value as the class itself?

a-matsuo replied:
You're right. I didn't notice it until you mentioned it.

mrossinek commented on qiskit_nature/problems/second_quantization/lattice/lattice_model_problem.py:
```diff
@@ -0,0 +1,116 @@
+        """
+        second_q_ops: ListOrDictType[SecondQuantizedOp] = self._lattice_model.second_q_ops()
+        if settings.dict_aux_operators:
+            second_q_ops = {"LatticeEnergy": second_q_ops}
```
Ideally, the internally used model should already return `second_q_ops` in accordance with
`settings.dict_aux_operators`. But this can also be fixed later as it requires deprecation.
Nonetheless, I suggest that you use `self._main_property_name` here in order to avoid the string
duplication.

mrossinek commented on qiskit_nature/problems/second_quantization/lattice/lattice_model_problem.py:
```diff
@@ -0,0 +1,116 @@
+            Callable[[int, Tuple[int, int]], List[Tuple[Tuple[int, ...], Tuple[int, ...]]]],
+        ] = "sd",
+    ) -> None:
+        """Generates the hopping operators and their commutativity information
```
More a question than a comment: do we need to duplicate this docstring or can Sphinx infer it from
the base class?

a-matsuo replied:
I've tested with my local environment. It looks Sphinx can infer this docstring from the base class (even I deleted this docstring in `LatticeModelProblem`, Sphinx could make a docstring from the base class for `LatticeModeProblem`'s `hopping_qeom_ops`).

woodsp-ibm replied:
It does inherit docstrings, we rely on this in many places so as to not have to repeat the same thing, rather just have it once. If there is some need, for some specific subclass behavior that needs explaining, then you can supply a docstring which will override any that might be inherited. But best avoided if possible.

a-matsuo replied:
Cool. I added a docstring to explicitly describe a `NotimplementedError`. Also, I added an error message for the `NotimplementedError`.

mrossinek reviewed (commented):
Besides these comments, this LGTM 👍

a-matsuo commented on qiskit_nature/problems/second_quantization/lattice/lattice_model_problem.py:
```diff
@@ -0,0 +1,116 @@
+        """
+        super().__init__()
+        self._lattice_model = lattice_model
+        self._main_property_name = "LatticeEnergy"
```
I talked this with @ikkoham. We can pass a `main property_name` to a constructor of `BaseProblem` to set it. e.g. `super().__init__("LatticeEnergy")`. The comment is legit since `BaseProblem` has `_main_property_name` variable with an empty string. 

What do you think? @mrossinek 
I can change all problem subclasses to set their name like `super().__init__("LatticeEnergy")` if you want. Then, I will open a new PR for that change.

mrossinek replied:
Sorry, I overlooked this comment before. Sounds like a reasonable change to me :+1: Could you do this in a separate PR?

woodsp-ibm commented on qiskit_nature/results/lattice_model_result.py:
```diff
@@ -0,0 +1,64 @@
+    def _formatted(self) -> List[str]:
+        """Formatted result as a list of strings"""
+        lines = []
+        lines.append("=== GROUND STATE ===")
+        lines.append(" ")
+        lines.append(
+            "* Lattice ground state energy " f": {np.round(self.computed_lattice_energies[0], 12)}"
+        )
```
Given this is now in a Problem , it can presumably be used with an excited state solver, such as using NumPyEigensolver with the ExcitedStatesEigensolver (and maybe later VQD) from Terra. Shouldn't this also format up any excited state energies that may be computed as presumably this is a valid use too right?

a-matsuo replied:
I copied messages for excited state energies from other problem classes.

woodsp-ibm commented on qiskit_nature/problems/second_quantization/lattice/lattice_model_problem.py:
```diff
@@ -115,4 +115,6 @@ def hopping_qeom_ops(
         """
-        raise NotImplementedError()
+        raise NotImplementedError(
+            "Currently, it's not implemented in the `LtticeProblemModel` class."
```
```suggestion
            "Currently, it's not implemented in the `LatticeProblemModel` class."
```
This copied the spelling error from above in the Returns docstring. I wonder whether we could say something like this problem cannot be used with QEOM solver. @mrossinek Any thoughts, otherwise the text says little beyond whats implied by the error itself.

mrossinek replied:
Looking at the code here as well as the qEOM code, it looks to me like this should generally be possible to implement. A quick test could be by testing the behavior of the `ExcitedStatesSolver` + `NumPyEigensolver` (not minimum) to see if that works as expected. If yes, I would handle that accordingly in the result printing and would open a separate issue to track the implementation of the `hopping_qeom_ops` method.

a-matsuo replied:
I changed `NotImplementedError` to return None. Then, `ExcitedStatesEigensolver` + `NumPyEigensolver` returns some values (I'm not sure whether this behavior is as expected or not).
<img width="509" alt="image" src="https://user-images.githubusercontent.com/47442626/169978856-622f7c06-d0a9-4735-9ec1-10da4388f78a.png">

woodsp-ibm replied:
> I changed NotImplementedError to return None. Then, ExcitedStatesEigensolver + NumPyEigensolver returns some values (I'm not sure whether this behavior is as expected or not).

I am not sure  I understand. Are you saying the excited states solver `ExcitedStatesEigensolver + NumPyEigensolver` fails if the `hopping_qeom_ops` raises an error - (and only if it did not, and returned None, did you get those values)? As far as I know its one called by QEOM solver and should never get invoked otherwise.

a-matsuo replied:
Sorry for the confusion. I thought it would fail and I hadn't check it with `hopping_qeom_ops` + `NotimplementedError`. However, I checked it works, even `hopping_qeom_ops` raises `NotimplementedError`.

a-matsuo commented:
I changed `get_default_filter_criterion` to explicitly return None in LatticeModelProblem. Can I just leave `get_default_filter_criterion` to return None?

mrossinek commented on qiskit_nature/problems/second_quantization/lattice/lattice_model_problem.py:
```diff
@@ -0,0 +1,125 @@
+                    methods, :meth:`generate_fermionic_excitations` or
+                    :meth:`generate_vibrational_excitations`.
+
+        Returns:
+            Currently, this function is not implemented in the `LtticeProblemModel` class and
+            always raises `NotImplementedError`.
```
Instead of `returns` this should say `Raises`. There is also still a typo here: `Lttice -> Lattice`.

mrossinek commented on qiskit_nature/problems/second_quantization/lattice/lattice_model_problem.py:
```diff
@@ -0,0 +1,125 @@
+            List[int],
+            Callable[[int, Tuple[int, int]], List[Tuple[Tuple[int, ...], Tuple[int, ...]]]],
+        ] = "sd",
+    ) -> None:
```
Despite this not being implemented, I would keep the return type hint unchanged for consistency (and also to document future expectation of what this method should return)

mrossinek commented on qiskit_nature/problems/second_quantization/lattice/lattice_model_problem.py:
```diff
@@ -0,0 +1,125 @@
+            "Currently, it's not implemented in the `LatticeProblemModel` class."
+        )
+
+    def get_default_filter_criterion(
+        self,
+    ) -> None:
```
Same here: I would try to keep the function definition aligned with the base class

mrossinek reviewed (approved):
This LGTM, thanks a lot @a-matsuo!

woodsp-ibm reviewed (approved):
LGTM too, thanks @a-matsuo
